### PR TITLE
Consolidate sdpattribute

### PR DESCRIPTION
**Pull request by Paul-E**

### CODE REVIEW

File: src/attribute_type.rs [hunk 1]
```diff
@@ -396,567 +307,593 @@ impl SdpAttributeSsrc {
...
 }
 
-#[derive(Clone)]
-pub struct SdpAttribute {
-    pub name: SdpAttributeType,
-    pub value: Option<SdpAttributeValue>,
+impl FromStr for SdpAttribute {
+    type Err = SdpParserError;
+
+    fn from_str(line: &str) -> Result<SdpAttribute, SdpParserError> {
```
Comment (nils-ohlmeier) [inline src/attribute_type.rs]: If I understand from_str() right it should be `Result<SdpAttribute, Self::Err>`. I guess without it you could declare two different things here.

File: src/attribute_type.rs [hunk 2]
```diff
@@ -396,567 +307,593 @@ impl SdpAttributeSsrc {
...
+    let mut ssrc = SdpAttributeSsrc::new(ssrc_id);
+    match tokens.next() {
+        None => (),
+        Some(x) => ssrc.set_attribute(x),
+    };
+    Ok(SdpAttribute::Ssrc(ssrc))
+}
+
+pub fn parse_attribute(value: &str) -> Result<SdpLine, SdpParserError> {
+    let attr: SdpAttribute = value.parse()?;
```
Comment (nils-ohlmeier) [inline src/attribute_type.rs]: I think we still want the `trim()` call to strip of white spaces. Could be either `value.trim().parse()` or you do the trimming inside of the `from_str()` function.

I'm also wondering if you could combine the two lines into a single line here.

File: src/attribute_type.rs [hunk 3]
```diff
@@ -396,567 +307,593 @@ impl SdpAttributeSsrc {
...
+
+    fn from_str(line: &str) -> Result<SdpAttribute, SdpParserError> {
+        let tokens: Vec<_> = line.splitn(2, ':').collect();
+        let name = tokens[0].to_lowercase();
+        let val = match tokens.get(1) {
+            Some(x) => x.trim(),
+            None => ""
+        };
+        if tokens.len() > 1 {
+            match name.as_str() {
```
Comment (nils-ohlmeier) [inline src/attribute_type.rs]: You probably need to add `to_lowercase()` here.

File: src/attribute_type.rs [hunk 4]
```diff
@@ -396,567 +307,593 @@ impl SdpAttributeSsrc {
...
+                    | "rtcp-mux" | "rtcp-rsize" | "sendonly"
+                    | "sendrecv" => {
+                        return Err(SdpParserError::Line{
+                            message: "This attribute is not allowed to have a value".to_string(),
+                            line: line.to_string()});
+                    },
+                _ => ()
+            }
+        }
+        match name.as_str() {
```
Comment (nils-ohlmeier) [inline src/attribute_type.rs]: Add `to_lowercase()` here as well.

File: src/attribute_type.rs [hunk 5]
```diff
@@ -396,567 +307,593 @@ impl SdpAttributeSsrc {
...
+            None => ""
+        };
+        if tokens.len() > 1 {
+            match name.as_str() {
+                "bundle-only" | "end-of-candidates" | "ice-lite"
+                    | "ice-mismatch" |  "inactive" |  "recvonly"
+                    | "rtcp-mux" | "rtcp-rsize" | "sendonly"
+                    | "sendrecv" => {
+                        return Err(SdpParserError::Line{
+                            message: "This attribute is not allowed to have a value".to_string(),
```
Comment (nils-ohlmeier) [inline src/attribute_type.rs]: I think it would be really useful to include the `name ` in the message as well.

File: src/attribute_type.rs [hunk 6]
```diff
@@ -396,567 +307,593 @@ impl SdpAttributeSsrc {
...
+                            line: line.to_string()});
+                    },
+                _ => ()
+            }
+        }
```
Comment (nils-ohlmeier) [inline src/attribute_type.rs]: I think before matching the `name` we should check here the `!val.is_empty()`. That would simplify `string_or_empty()` and also ensure that all other parsers are guaranteed to have a value.

Comment (Paul-E) [inline src/attribute_type.rs]: Must all attributes that have a value have a nonempty value? In a lot of the parsing methods a value is checked for, and the error message is more specific.

File: src/attribute_type.rs (hunk 3 above)
Comment (Paul-E) [inline src/attribute_type.rs]: Which line are you referencing?

**Comment by Paul-E** [comment]
is the from_str approach going to conflict with the new error handling approach?

File: src/attribute_type.rs (hunk 4 above)
Comment (Paul-E) [inline src/attribute_type.rs]: Name should already be lowercase if you look at line 356

File: src/attribute_type.rs (hunk 3 above)
Comment (nils-ohlmeier) [inline src/attribute_type.rs]: I meant the `name.as_str()` call.

Comment (nils-ohlmeier) [inline src/attribute_type.rs]: Never mind. I did not see the one in line 356

File: src/attribute_type.rs (hunk 6 above)
Comment (nils-ohlmeier) [inline src/attribute_type.rs]: Oh do you mean unknown attributes? That's right we don't know what to expect from an unknown attribute.
Alright I guess that kills that idea/comment.

**Pull request merged**